### DISCUSSION TITLE
Mailgun emails not being triggered in production. (Unauthorized Error)

### DISCUSSION THREAD

**Discussion by srestraj** [vercel/next.js#64607] [Help]
### Summary

I have a Next.js 14 project where I've integrated the Mailgun API to send emails. The integration works perfectly fine when I'm running the project locally. However, after deploying to production on Vercel, I'm consistently encountering an "Unauthorized" error when attempting to send emails via the Mailgun API.

I know that this "Unauthorized" error is being thrown by the Mailgun API, but we've already added Vercel's IP, i.e. 76.76.21.21 to our domain's whitelist on Mailgun, but still no luck.

I've thoroughly checked and verified the API keys, environment variables and permissions and they're all fine. I've also included my code below for reference.

```tsx
import { NextRequest, NextResponse } from "next/server";
import FormData from "form-data";
import Mailgun from "mailgun.js";

const { MAILGUN_EMAIL, MAILGUN_TO_EMAIL, MAILGUN_API, MAILGUN_DOMAIN } =
  process.env;

export async function POST(request: NextRequest, res: NextResponse) {
  const { name, email, ticketCount, attendingAs, exhibitorDetails } =
    await request.json();

  const mailgun = new Mailgun(FormData);

  try {
    const mg = mailgun.client({
      username: "api",
      key: MAILGUN_API as string,
    });

    const mailData = {
      from: MAILGUN_EMAIL,
      to: MAILGUN_TO_EMAIL,
      subject: `${name} just signed up for the Expo`,
      text: `${name} just signed up for ${ticketCount} tickets as ${attendingAs} for the expo with email: ${email}.`,
      html: `
      <body>
        <div style="background: #f7f7f7; padding: 20px 0; margin:0">
           body content here
        </div>
      </body>`,
    };

    await new Promise((resolve, reject) => {
      mg.messages
        .create(MAILGUN_DOMAIN as string, mailData)
        .then((message) => {
          resolve(message);
        })
        .catch((error) => {
          reject(error);
          return NextResponse.json({ error: error }, { status: error.status });
        });
    });
    return NextResponse.json(
      {
        message:
          "Thank you for your interest in attending the expo. We hope to see you there.",
      },
      { status: 200 }
    );
  } catch (error: any) {
    return NextResponse.json({ error: error }, { status: error.status });
  }
}
```
Could someone please assist me in troubleshooting this issue? Any insights or guidance would be greatly appreciated.

### Example

https://www.mewingapp.co/mewing-expo

**Comment by icyJoseph** [comment]
In the old Next.js docs this line was included:

> Note: In order to keep server-only secrets safe, Next.js replaces process.env.* with the correct values at build time. This means that process.env is not a standard JavaScript object, so you’re not able to use object destructuring.

So this won't work:

```ts
const { MAILGUN_EMAIL, MAILGUN_TO_EMAIL, MAILGUN_API, MAILGUN_DOMAIN } =
  process.env;
```

I guess, this happens locally too if you do `next build`, on your local system.

Also you have the wrong function signature:

```ts
import { NextRequest, NextResponse } from "next/server";
import FormData from "form-data";
import Mailgun from "mailgun.js";

const MAILGUN_API = process.env.MAILGUN_API; // same for the others
// process.env is not a standard JavaScript object, so you can't just destructure it

export async function POST(request: NextRequest) {
  const { name, email, ticketCount, attendingAs, exhibitorDetails } =
    await request.json();

  const mailgun = new Mailgun(FormData);

  try {
    if (!MAILGUN_API) {
        throw new Error("MAILGUN_API is not set")
    }
   // same for the others

    const mg = mailgun.client({
      username: "api",
      key: MAILGUN_API,
    });

    const mailData = {
      from: MAILGUN_EMAIL,
      to: MAILGUN_TO_EMAIL,
      subject: `${name} just signed up for the Expo`,
      text: `${name} just signed up for ${ticketCount} tickets as ${attendingAs} for the expo with email: ${email}.`,
      html: `
      <body>
        <div style="background: #f7f7f7; padding: 20px 0; margin:0">
           body content here
        </div>
      </body>`,
    };

    await new Promise((resolve, reject) => {
      mg.messages
        .create(MAILGUN_DOMAIN as string, mailData)
        .then((message) => {
          resolve(message);
        })
    });
    return NextResponse.json(
      {
        message:
          "Thank you for your interest in attending the expo. We hope to see you there.",
      },
      { status: 200 }
    );
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : "unknown error"
    return NextResponse.json({ error: errorMessage }, { status: error.status });
  }
}
```

**Reply by srestraj** [reply]
Thank you for your response. But, even with the suggested changes, I have the 401 Unauthorized Error in the Mailgun API.

![image](https://github.com/vercel/next.js/assets/25652144/f01e7cfd-a4a8-420f-bfdd-18601ebda67e)

I know we have to add some IPs to Mailgun's **IP Allowlist** to resolve the 401 error. For now, we've added Vercel's `76.76.21.21` to that list since we're using Vercel for the deployment and hosting. But we're still getting the 401 error. Any chance you might know what we should put there instead of `76.76.21.21` or any changes we might need to do there? On DNS lookup for our website, we also get `76.76.21.22` and `76.76.21.61` as IPs. Should we also add these to the list?

**Reply by Techstudiosbe** [reply]
HI did u find the solution ?

**Reply by srestraj** [reply]
Hi @Techstudiosbe, so I came across one Vercel article, which suggested that on the free plan, we don't have the option to create custom IP for our domains as Vercel deployments use a range of IP addresses that can't be easily specified for Mialgun's IP Allowlist.

In our case, we decided to change our hosting provider to Hostinger since our team didn't want to switch the email providers. If you can switch to a different email provider, you could try going for Zoho for your emails instead of Mailgun. It's much easier to setup and works perfectly fine with Vercel too, without any errors.
